LoraBraveSensor.ino

```diff
@@ -7,46 +7,49 @@
 #include "systemTimers.h"
 #include "main.h"
 
-#define DOOR_SENSOR_PIN PA7
-#define MOTION_SENSOR_PIN PA6
 #define BATTERY_PIN PA5
 
-DoorSensor doorSensor = DoorSensor(DOOR_SENSOR_PIN);
-MotionSensor motionSensor = MotionSensor(MOTION_SENSOR_PIN);
 AnalogSensor battery = AnalogSensor(BATTERY_PIN);
 
 void setup()
 {
   lora::setupOTAA();
   fsm::setupFSM();
   attachInterrupt(
-      digitalPinToInterrupt(DOOR_SENSOR_PIN), [] {}, CHANGE);
+      digitalPinToInterrupt(DOOR_SENSOR_PIN), fsm::addToSensorDataQueue, CHANGE);
   attachInterrupt(
-      digitalPinToInterrupt(MOTION_SENSOR_PIN), [] {}, CHANGE);
+      digitalPinToInterrupt(MOTION_SENSOR_PIN), fsm::addToSensorDataQueue, CHANGE);
   lora::sendUplink("Modify the comment if this gets sent :3"); // No idea why first message always fails to send
+  fsm::addToSensorDataQueue(); 
 }
 
 void loop()
 {
-  int stateSleepTimer = fsm::stateHandler(doorSensor, motionSensor);
+  int stateSleepTimer = fsm::handleState(); 
   int heartBeatTimer = heartbeat::getRemainingDuration();
-  DEBUG_SERIAL_LOG_MORE.printf("State sleep timer: %is\r\n", stateSleepTimer / 1000);
-  DEBUG_SERIAL_LOG_MORE.printf("Heartbeat timer: %is\r\n", heartBeatTimer / 1000);
-  int sleepDuration; 
-  // Logic nessessary as timer == 0 equates infinite sleep, which is greater than any positive timer value
-  if (stateSleepTimer == 0) {
-    sleepDuration = heartBeatTimer;
-  } else if (heartBeatTimer == 0) {
-    sleepDuration = stateSleepTimer;
-  } else {
-    sleepDuration = min(stateSleepTimer, heartBeatTimer);
-  }
-  DEBUG_SERIAL_LOG.printf("Sleeping for %is\r\n", sleepDuration / 1000);
+  int uplinkTimer = lora::getRemainingDuration(); 
+  DEBUG_SERIAL_LOG.printf("State sleep timer: %is, ", stateSleepTimer / 1000);
+  DEBUG_SERIAL_LOG.printf("Heartbeat timer: %is, ", heartBeatTimer / 1000);
+  DEBUG_SERIAL_LOG.printf("Uplink timer: %is\r\n", uplinkTimer / 1000);
+
+  int sleepDuration = min(stateSleepTimer, min(heartBeatTimer, uplinkTimer));
+  DEBUG_SERIAL_LOG.printf("Attempting to sleep for %is...", sleepDuration / 1000);
+  
   if (sleepDuration > 0)
   {
-    api.system.timer.create((RAK_TIMER_ID)FSM_TIMER, (RAK_TIMER_HANDLER)[](void *){}, RAK_TIMER_ONESHOT);
+    api.system.timer.create((RAK_TIMER_ID)FSM_TIMER, (RAK_TIMER_HANDLER)fsm::addToSensorDataQueue, RAK_TIMER_ONESHOT);
     api.system.timer.start((RAK_TIMER_ID)FSM_TIMER, sleepDuration, (void *)1);
   }
-  api.system.sleep.all();
+  if (fsm::stateQueueEmpty()) {
+    DEBUG_SERIAL_LOG.println("State queue empty, sleeping");
+    if (lora::isUplinkProcess()) {
+      delay(sleepDuration); 
```

Comment:
Sounds like the use of the `delay` function is discouraged because the program won't do anything else while it's waiting: 

https://www.arduino.cc/reference/en/language/functions/time/delay/
http://gammon.com.au/interrupts

Reply:
Previously, I had the blocking delays in the lora code. I have reverted it to that, as I believe it improves uplink consistency.

Reply:
We're unsure why this is advised against. So we're still using it.